node/src/reactor/participating.rs

```diff
@@ -476,10 +476,18 @@ impl reactor::Reactor for Reactor {
             init_consensus_effects,
         ));
 
-        contract_runtime.set_initial_state(maybe_latest_block_header.map_or_else(
-            || chainspec_loader.initial_execution_pre_state(),
-            |latest_block_header| ExecutionPreState::from(&latest_block_header),
-        ));
+        let execution_pre_state = match maybe_latest_block_header {
+            Some(latest_block_header)
+                if chainspec_loader
+                    .initial_execution_pre_state()
+                    .next_block_height()
+                    <= latest_block_header.height() =>
```

Comment:
I think there should be a `chainspec_loader.after_upgrade()` endpoint now that returns `true` only if we've just gone through an upgrade and there are no more blocks in the storage.

Reply:
I think that would require us to store more state in the chainspec loader, wouldn't it? We'd have to store the "we've just gone through an upgrade in this run" bit of information. And then this `after_upgrade()` method would have to send a request to storage, so it would complicate this part of code quite significantly (one more effect-event round-trip).

I'm not totally opposed to the idea, but I think it's probably more complicated than it might look at the first glance.

Reply:
> I think that would require us to store more state in the chainspec loader, wouldn't it? We'd have to store the "we've just gone through an upgrade in this run" bit of information. 

I meant that we already do that - there already is a `after_upgrade()` method in the chainspec loader and it does read the storage to determine it.